### DISCUSSION TITLE
How to save the results of multi-class segmentation？

### DISCUSSION THREAD

**Discussion by Alex8211101** [Project-MONAI/MONAI#4280] [Q&A] [answered]
I'm a beginner in monai and recently working on a 3D medical image segmentation project. I am referring to the spleen segmentation project, but my images have four classes, and when I save the segmentation results to nifti files, I find that only the background is displayed. How to save segmentation results of different classes to nifti files?
![微信图片_20220516201804](https://user-images.githubusercontent.com/76839078/168590952-7b285ccd-d745-439d-827f-ec12ef2ed8bc.jpg)

**Comment by Alex8211101** [comment]
![微信图片_20220516204057](https://user-images.githubusercontent.com/76839078/168594618-04de4e11-bb27-4733-9b42-323682a21539.jpg)

**Comment by Nic-Ma** [comment]
Hi @Alex8211101 ,

Thanks for your interest here.
I think maybe you can refer to the Brats tutorial which also contains multiple classes: https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/brats_segmentation_3d.ipynb

Thanks.

**Reply by Alex8211101** [reply]
Thanks for your reply, maybe my question is not clear. I can already calculate the dice metrics of different organs and display the slice segmentation results, but when saving, only the background can be saved, and other organs cannot be saved in nifti format, as shown in the 3d slicer software.

**Comment by rijobro** [comment, marked as answer]
Looks like your image will contain 4 channels, and these will be one-hotted. This means that you should likely be using the `Softmax` activation, not `Sigmoid`. From here, you can then do `argmax` to convert from a one-hotted image into an image with a single channel filled with integers, each integer corresponding to a different label. I suspect that once you save this, it will look as you expect.

Edit: sorry, I didn't check your code properly before replying -- you already have the `argmax`. Have you checked that your network is doing what you expect? You could do this with a `matplotlib.imshow`. If it looks normal in `imshow`, then maybe the problem is just the Slicer viewer settings. For debugging you could also remove the `Invertd` just in case that's causing any issues (although I doubt it would be doing any inverting of the labelfields).

**Reply by rijobro** [reply]
After discretising and argmax-ing, you could do something like (untested):

```python
[np.where(im == x).sum() for x in np.unique(im)]
```

to extract the number of voxels in each label. You would probably expect the 0th label (background) to have the largest fraction.

**Reply by Alex8211101** [reply]
Thank you very much for your reply, I tried using `SplitChanneled` and got the desired results, I will also use your suggestion to achieve it, thank you again for your suggestion.
